Clean up comments in nestkernel

jessica-mitchell opened the pull request:
This PR aims to make the comments (doxygen and c++) more consistent and useful.
So far, only a few files have been touched because I would like some feedback as to whether I'm going in the right direction before proceeding. @jougs could you take a look?

Changes implemented:

* C++ comments single line (max 2 lines if sentence is long): // 
* C++ comments multile line: /* * * * */
* doxygen comments single line (max 2 lines if sentence is long): //!
* doxygen comments multiline: /** * * * * */
* doxygen comments side: //!<
* avoid doxygen comments in .cpp files, move to header file 
* avoid duplicating code in comments; only include code(parameters, functions etc) that also have additional context,  remove redundancies
* include the variable name in functions in header file to match cpp file.

jougs commented on nestkernel/conn_builder.h:
```diff
@@ -105,7 +109,19 @@ class ConnBuilder
+ * @param update amount of connected synaptic elements to update
+ * @return
+ */
+  bool change_connected_synaptic_elements( index snode_id, index tnode_id, const int thread tid, int update );
```
This should fix the build problems:

```suggestion
  bool change_connected_synaptic_elements( index snode_id, index tnode_id, const int tid, int update );
```

jougs reviewed (commented):
Thanks for making the chaos and invonsistencies visible. I'm reviewing in greater detail once the crying stops. For now, I have a couple of generic comments:

* I prefer the following doxygen style for functions and classes, even if the comment itself is only a single line:
  ```
  /**
   * ...
   */
  ```
  (I also think this is the predominant style presently used, so it will save you a lot of trouble down the line :smile:)
* I would use `//!` for long one-liners above variables and `//!<` for short ones behind them

Maybe we should set up a video session where we go through the [Doxygen documentation](https://www.doxygen.nl/manual/docblocks.html) and ask people what they (dis)like most?

jougs commented on nestkernel/clopath_archiving_node.cpp:
```diff
@@ -64,13 +64,15 @@ nest::ClopathArchivingNode::ClopathArchivingNode( const ClopathArchivingNode& n
+   * Since we write into the buffer before we read from it, we have to
+   * add 1 to the size of the buffers.
+   */
+  layed_u_bars_idx_ = 0;
```
Here's another one for fixing the build
```suggestion
  delayed_u_bars_idx_ = 0;
```

jessica-mitchell commented:
> Thanks for making the chaos and invonsistencies visible. I'm reviewing in greater detail once the crying stops. For now, I have a couple of generic comments:
> 
>     * I prefer the following doxygen style for functions and classes, even if the comment itself is only a single line:
>       ```
>       /**
>        * ...
>        */
>       ```
Sure, I think you're right that it's more common
> 
>     * I would use `//!` for long one-liners above variables and `//!<` for short ones behind them

When you say above variables, can you give me an example of the syntax you'd expect this kind of comment? Where would not want this comment style? (it's just easier if I have an example to look up, because it's easy to get confused staring at slashes and askterisks all day ;) )
> 
> Maybe we should set up a video session where we go through the [Doxygen documentation](https://www.doxygen.nl/manual/docblocks.html) and ask people what they (dis)like most?

Yah sure, I think that could work.

jougs commented:
Here are some examples that hopefully clarify my style preferences mentioned above:
```c++
/**
 * Set the minimal and maximal delay and override automatically determined values.
 */
void set_delay_extrema(delay min_delay, delay, max_delay();

//! The variable min_delay holds the value for the smallest transmission delay in the network in steps
delay min_delay;

delay max_delay //!< The largest transmission delay in the network (steps)
```

In real code, I would then actually prefer if similar variables were documented consistently, i.e. both `min_delay` and `max_delay` with either the long version, or with the short one, depending on the length of the comment.

heplesser commented on doc/devdoc/cppcomments.rst:
```diff
@@ -0,0 +1,40 @@
+============================
+
+* C++ comments single line (max 2 lines if sentence is long): //
+* C++ comments multile line: ``/* * * * */``
```
This looks a bit confusing. I'd typeset it over multiple lines.

heplesser commented on doc/devdoc/cppcomments.rst:
```diff
@@ -0,0 +1,40 @@
+Functions and classes
+---------------------
+
+doxygen style: ``/** * */``
```
I would typeset this over multiple lines as well.

jessica-mitchell replied:
done

heplesser commented on doc/devdoc/cppcomments.rst:
```diff
@@ -0,0 +1,40 @@
+---------
+
+
+doxygen comments single line (max 2 lines if sentence is long): //!
```
At the beginning of sentences, I'd capitalize in keeping with the doxygen website.
```suggestion
Doxygen comments single line (max 2 lines if sentence is long): //!
```

Should there be backticks around //!, and similar below?

jessica-mitchell replied:
yes

heplesser commented on doc/devdoc/cppcomments.rst:
```diff
@@ -0,0 +1,40 @@
+Rules for C++  code comments
+============================
+
```
Maybe add a link to doxygen documentation?

jessica-mitchell replied:
done

heplesser commented on doc/devdoc/cppcomments.rst:
```diff
@@ -0,0 +1,40 @@
+Rules for C++  code comments
+============================
```
```suggestion
Rules for C++ code comments
===========================
```

heplesser commented on nestkernel/archiving_node.h:
```diff
@@ -50,44 +50,32 @@ class ArchivingNode : public StructuralPlasticityNode
 public:
   /**
-   * \fn ArchivingNode()
-   * Constructor.
+   * Constructor
    */
```
Trivial comments like this on constructor, copy constuctor or assignment operator provide no information. I would remove them completely.

jessica-mitchell replied:
Removed these, also removed destructor comments as well

heplesser commented on nestkernel/archiving_node.h:
```diff
@@ -50,44 +50,32 @@ class ArchivingNode : public StructuralPlasticityNode
   /**
-   * \fn double get_K_value(long t)
-   * return the Kminus (synaptic trace) value at t (in ms). When the trace is
+   * Return the Kminus (synaptic trace) value at t (in ms). When the trace is
```
If I am not entirely mistaken, the same rules apply to doxygen comments as to Python docstrings: the first line should give a short summary of what a function does, followed by an empty line and then, if relevant, details. So here 
```c++
/**
 * Return the Kminus (synaptic trace) value at t (in ms). 
 * 
 * When the trace is ...
```

jessica-mitchell replied:
Added a new line after first line description in comments

heplesser commented on nestkernel/archiving_node.h:
```diff
@@ -130,26 +114,25 @@ class ArchivingNode : public StructuralPlasticityNode
-  // number of incoming connections from stdp connectors.
-  // needed to determine, if every incoming connection has
-  // read the spikehistory for a given point in time
+  /*
+   * number of incoming connections from stdp connectors.
+   *  needed to determine, if every incoming connection has
+   *  read the spikehistory for a given point in time
+   */
```
```suggestion
  /*
   * Number of incoming connections from stdp connectors.
   *
   * It is needed to determine if every incoming connection has
   * read the spike history for a given point in time.
   */
```

heplesser commented on nestkernel/clopath_archiving_node.cpp:
```diff
@@ -170,9 +172,11 @@ nest::ClopathArchivingNode::get_LTP_history( double t1,
-    // To have a well defined discretization of the integral, we make sure
-    // that we exclude the entry at t1 but include the one at t2 by subtracting
-    // a small number so that runner->t_ is never equal to t1 or t2.
+    /*
```
I'd put an empty line above this comment.

heplesser reviewed (changes requested):
@jessica-mitchell I added some comments to the guidelines as two examples and would like your and @jougs' feedback before commenting more.

terhorstd commented:
@jessica-mitchell, do you have a status update on this one?

jougs commented:
@jessica-mitchell: What is the status of this? Thanks!

jessica-mitchell commented on doc/htmldoc/developer_space/cppcomments.rst:
```diff
@@ -0,0 +1,80 @@
+Guidelines for C++  code comments
```
```suggestion
Guidelines for C++ code comments
```

jessica-mitchell commented on nestkernel/conn_builder.h:
```diff
@@ -308,10 +353,42 @@ class AllToAllBuilder : public ConnBuilder
   }
 
 protected:
+<<<<<<< HEAD
```
```suggestion
```

jessica-mitchell commented on doc/htmldoc/developer_space/cppcomments.rst:
```diff
@@ -0,0 +1,80 @@
+..  Include the variable name in functions in header file to match cpp file.
+
+
+Generate HTML do doxygen comments
```
```suggestion
Generate HTML from doxygen comments
```

jessica-mitchell commented on doc/htmldoc/developer_space/cppcomments.rst:
```diff
@@ -0,0 +1,80 @@
+
+
+Generate HTML do doxygen comments
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

jessica-mitchell commented:
@jougs @heplesser I think I addressed all the comments so far. In addition to removing the Constructor comments, I also removed Destructor.

I cleaned up the guidelines (doc/htmldoc/developer_space/cppcoments.rst) to be more clear, but I am unsure of how to describe when to use C++ comment style `/* */` versus doxygen style `/** */`.

jougs commented on nestkernel/connection.h:
```diff
@@ -319,22 +322,28 @@ Connection< targetidentifierT >::check_connection_( Node& dummy_target,
+   * this line might throw an exception
+   */
   source.send_test_event( dummy_target, receptor_type, get_syn_id(), true );
 
-  // 2. does the target accept the event type sent by source
-  // try to send event from source to target
-  // this returns the port of the incoming connection
-  // p must be stored in the base class connection
-  // this line might throw an exception
+  /*
+   * 2. does the target accept the event type sent by source
+   * try to send event from source to target
+   * this returns the port of the incoming connection
+   * p must be stored in the base class connection
+   * this line might throw an exception
+   */
   target_.set_rport( source.send_test_event( target, receptor_type, get_syn_id(), false ) );
 
-  // 3. do the events sent by source mean the same thing as they are
-  // interpreted in target?
-  // note that we here use a bitwise and operation (&), because we interpret
-  // each bit in the signal type as a collection of individual flags
+  /*
+   * 3. do the events sent by source mean the same thing as they are
+   * interpreted in target?
+   * note that we here use a bitwise and operation (&), because we interpret
+   * each bit in the signal type as a collection of individual flags
+   */
```
Revert

jougs commented on nestkernel/connection_id.h:
```diff
@@ -106,4 +106,4 @@ ConnectionID::get_port() const
 } // namespace
 
-#endif /* #ifndef CONNECTION_ID_H */
+#endif // #ifndef CONNECTION_ID_H
```
Revert

jougs reviewed (commented):
Here's a bunch of comments. Some of them are rather generic and should be applied in many places. Probably something like `git difftool --dir-diff upstream/master` could help with bulk reverting changes more easily than purely by hand. I will have another look once my comments are addressed.

jessica-mitchell commented on nestkernel/common_properties_hom_w.h:
```diff
@@ -24,17 +24,17 @@
-
-   SeeAlso: static_synapse
-*/
+ *
+ *  Name: static_synapse_hom_w - Static synapse type
+ *       using homogeneous weight, i.e. all synapses
+ *        feature the same w.
+ *
+ *  FirstVersion: April 2008
+ *
+ *  Author: Moritz Helias, Susanne Kunkel
+ *
+ *  SeeAlso: synapsedict, static_synapse
+ */
```
I would propose that a separate PR be made to deal with any of  these SLI documentation sections that needs to be converted into reStructured Text

jougs replied:
I would not bother with any SLI things anymore at this point as these comments will anyway just be deleted in PyNEST NG together with all of SLI.

jessica-mitchell commented on nestkernel/conn_builder_conngen.h:
```diff
@@ -61,6 +117,6 @@ class ConnectionGeneratorBuilder : public ConnBuilder
+#endif // ifdef HAVE_LIBNEUROSIM
 
-#endif /* ifdef CONN_BUILDER_CONNGEN_H */
+#endif // ifdef CONN_BUILDER_CONNGEN_H
```
done

jessica-mitchell commented on nestkernel/conn_builder_conngen.cpp:
```diff
@@ -349,4 +295,4 @@ ConnectionGeneratorBuilder::cg_get_ranges( RangeSet& ranges, const NodeCollectio
 } // namespace nest
 
-#endif /* ifdef HAVE_LIBNEUROSIM */
+#endif /* ifdef HAVE_LIBNEUROSIM  */ */
```
```suggestion
#endif /* ifdef HAVE_LIBNEUROSIM  */
```

jessica-mitchell commented on nestkernel/deprecation_warning.cpp:
```diff
@@ -58,4 +58,4 @@ DeprecationWarning::deprecation_warning( std::string name, std::string new_name
 }
 
-#endif /* DEPRECATION_WARNING_IMPL_H */
+#endif /* DEPRECATION_WARNING_IMPL_H  */ */
```
```suggestion
#endif /* DEPRECATION_WARNING_IMPL_H  */
```

jessica-mitchell commented on nestkernel/mpi_manager.cpp:
```diff
@@ -95,7 +95,7 @@ nest::MPIManager::init_mpi( int*, char*** )
 }
 
-#else /* HAVE_MPI */
+#else // HAVE_MPI
```
```suggestion
#else /* HAVE_MPI */
```

jessica-mitchell commented on nestkernel/mpi_manager.cpp:
```diff
@@ -124,11 +124,11 @@ nest::MPIManager::init_mpi( int* argc, char** argv[] )
     // get a communicator from MUSIC
     set_communicator( static_cast< MPI_Comm >( kernel().music_manager.communicator() ) );
-#else  /* #ifdef HAVE_MUSIC */
+#else  // #ifdef HAVE_MUSIC
```
```suggestion
#else  /* #ifdef HAVE_MUSIC */
```

jessica-mitchell commented on nestkernel/mpi_manager.cpp:
```diff
@@ -124,11 +124,11 @@ nest::MPIManager::init_mpi( int* argc, char** argv[] )
     MPI_Init_thread( argc, argv, MPI_THREAD_FUNNELED, &provided_thread_level );
     set_communicator( MPI_COMM_WORLD );
-#endif /* #ifdef HAVE_MUSIC */
+#endif /* #ifdef HAVE_MUSIC/ */
```
```suggestion
#endif /* #ifdef HAVE_MUSIC */
```

jessica-mitchell commented on nestkernel/mpi_manager.cpp:
```diff
@@ -276,7 +276,7 @@ nest::MPIManager::mpi_finalize( int exitcode )
 }
 
-#else /* #ifdef HAVE_MPI */
+#else // #ifdef HAVE_MPI
```
```suggestion
#else /* #ifdef HAVE_MPI */
```

jessica-mitchell commented on nestkernel/mpi_manager.cpp:
```diff
@@ -982,11 +978,8 @@ nest::MPIManager::time_communicate_alltoallv( int num_bytes, int samples )
 }
 
-#else /* #ifdef HAVE_MPI */
+#else // #ifdef
```
```suggestion
#else /* #ifdef */
```

jessica-mitchell commented on nestkernel/mpi_manager.cpp:
```diff
@@ -1100,4 +1090,4 @@ nest::MPIManager::communicate_recv_counts_secondary_events()
 }
 
-#endif /* #ifdef HAVE_MPI */
+#endif /* #ifdef HAVE_MPI  */ */
```
```suggestion
#endif /* #ifdef HAVE_MPI  */
```

jessica-mitchell commented on nestkernel/mpi_manager.h:
```diff
@@ -755,7 +768,7 @@ MPIManager::time_communicate_alltoallv( int, int )
 }
 
-#endif /* HAVE_MPI */
+#endif /* HAVE_MPI  */ */
```
```suggestion
#endif /* HAVE_MPI  */
```

jessica-mitchell commented on nestkernel/mpi_manager_impl.h:
```diff
@@ -46,20 +46,18 @@ nest::MPIManager::get_process_id_of_vp( const thread vp ) const
 #ifdef HAVE_MUSIC
 extern MPI::Intracomm comm;
-#else  /* #ifdef HAVE_MUSIC */
+#else  // * #ifdef HAVE_MUSIC
```
```suggestion
#else  /* #ifdef HAVE_MUSIC */
```

jessica-mitchell commented on nestkernel/music_manager.cpp:
```diff
@@ -97,7 +97,7 @@ MUSICManager::enter_runtime( double h_min_delay )
 }
 
-#else /* #ifdef HAVE_MUSIC */
+#else // #ifdef HAVE_MUSIC
```
```suggestion
#else /* #ifdef HAVE_MUSIC */
```

jessica-mitchell commented on nestkernel/music_manager.cpp:
```diff
@@ -124,7 +124,7 @@ MUSICManager::music_finalize()
   music_runtime->finalize();
   delete music_runtime;
-#else /* #ifdef HAVE_MUSIC */
+#else // #ifdef HAVE_MUSIC
```
```suggestion
#else /* #ifdef HAVE_MUSIC */
```

jessica-mitchell commented:
@jougs the conflict is fixed

jessica-mitchell commented:
@jougs ping!

jougs commented:
@jessica-mitchell: [pong!](https://github.com/jessica-mitchell/nest-simulator/pull/19)

jougs reviewed (approved):
Many thanks for this heroic effort and your patience. Merging without second review due to discussion in several meetings and because this "only" touches documentation.
